proposals/awslogs.md

```diff
@@ -0,0 +1,195 @@
+4. User wants to monitor their task/service, so they continually stream the logs.
+
+## Phase 1 Solution
+Top level `ecs-cli logs` command that will not use the docker compose file. This allows it to be used by a wide array of ECS customers, not just compose users. The command will allow customers to find logs for a given task.
+
+```
+ecs-cli logs --help
+--follow                         Stream logs (continuously poll for updates)
+--task-id                        [Required] View logs for a given Task ID
+--task-def                      Required with Task ID if the task has been stopped already. Format: family:revision
+--pattern                        Substring to search for within the logs.
+--container-name, -c     Filter logs for a given container definition
+--since                           Filter logs in the last X minutes (can not be used with start time and end time)
+--start-time                    Filter logs within a time frame, use with --end-time
+--end-time                     Filter logs within a time frame, use with --end-time
+--time-stamps, -t           View time-stamps with the logs
+```
+
+```
+ecs-cli logs --task-id d86079d1-6858-45e9-8ce2-1ba881c55c12 --time-stamps
+Time-stamp            Message
+2017-09-28 22:32:11 WordPress not found in /var/www/html - copying now...
+2017-09-28 22:32:11 Complete! WordPress has been successfully copied to /var/www/html
+2017-09-28 22:32:12 AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 172.17.0.3. Set the 'ServerName' directive globally to suppress this message
+2017-09-28 22:32:12 AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 172.17.0.3. Set the 'ServerName' directive globally to suppress this message
+2017-09-28 22:32:12 [Wed Sep 27 22:32:12.300422 2017] [mpm_prefork:notice] [pid 1] AH00163: Apache/2.4.10 (Debian) PHP/5.6.31 configured -- resuming normal operations
+2017-09-28 22:32:12 [Wed Sep 27 22:32:12.300456 2017] [core:notice] [pid 1] AH00094: Command line: 'apache2 -D FOREGROUND'
+```
+
+### Implementation
+
+- The logs implementation will not include any pagination- the command will return all logs corresponding to the specified search. We expect most users will be piping the output of the command to save it to a file, so this should not be a problem.
+- If the user has not specified a log stream prefix in their task definition, then the command will fail and print an error message. Because without the log stream prefix set, we have no way of getting the logs for an individual task.
+- For performance reasons, the command will only pull from *a single log group*. If the customer has not configured all of their container definitions to use the same log group, then the command will fail with an error and tell the customer they must re-run the command with the `--container-name` argument. This way, only 1 log group needs to be queried.
```

Comment:
I think this subtly encourages some behavior that we don't really want to encourage.  Log groups should be homogeneous in terms of the log format/content type, since the filtering and alarming features of CloudWatch Logs work on a log group basis.  By failing and printing a message saying you need to specify the container name, this encourages people to send all their container logs to the same log group; sending all the logs to the same log group ultimately makes it harder to use the filtering and alarming features.  I think we could either always require the container name to be specified (regardless of whether multiple containers share the same log group) or just take whatever performance hit it entails to actually query all the log groups.

Reply:
Hmm, so the issue with querying multiple log groups is that each log group will entail a separate call (actually a series of paginated calls). That's not too bad, the issue is that when the output is printed, the log events will not be ordered by time- unless we do some very expensive in memory sorting of the log events. We should may be discuss this further; when I was designing this feature, the feedback I got was that the log events should always be printed ordered by time, and that the command mustn't be too slow. Hence I made this design decision.